### DISCUSSION TITLE
Problem with callable type with generic

### DISCUSSION THREAD

**Discussion by felinae98** [microsoft/pyright#4884] [Q&A] [answered]
In my scenario, I need a function to operate lists of different types (list is just an example).

If I define the function type in generic alias as follows
```python
from typing import Callable, TypeVar


T = TypeVar('T')
Operation = Callable[[list[T]], list[T]]

def my_func(f: Operation):
# def my_func(f: Callable[[list[T]], list[T]])
    a: list[int] = []
    aa = f(a)
    b: list[str] = []
    bb = f(b)
```
![image](https://user-images.githubusercontent.com/23295345/229268289-5ffa629f-9b94-4298-ad5d-24594a0bd9a2.png)

The type of aa and bb are inferred as list[Unknow], which should be list[int] and list[str].

If I define the function type in function signature, an error just occurred:
```python
from typing import Callable, TypeVar


T = TypeVar('T')
Operation = Callable[[list[T]], list[T]]

# def my_func(f: Operation):
def my_func(f: Callable[[list[T]], list[T]]):
    a: list[int] = []
    aa = f(a)
    b: list[str] = []
    bb = f(b)
```
![image](https://user-images.githubusercontent.com/23295345/229268363-afe6aec4-024b-4f21-a387-05d6f84e600d.png)

I wonder how to deal with that problem.

I'm using pyright 1.1.301 installed as a coc.nvim plugin

**Comment by erictraut** [comment, marked as answer]
You've defined a generic alias that takes a single type argument, but you haven't supplied any type argument when you use it. That means a type checker will assume the type argument should be `Any` (or `Unknown`, which is an "implied" `Any`).

If you want to parameterize the type alias with a function-scoped TypeVar, you can do the following:

```python
def my_func(f: Operation[T]):
    ...
```
